Add expected installer return codes

lechacon opened the pull request:
Adding expected return codes for installer. Expected return codes are translated to specific HResults when found and show a custom error message.
* Updated v1.1 manifest schema with new field for expected return codes. Updated file would look like this:
```
Installers: 
    - InstallerUrl: https://example.com
      InstallerType: exe
      ExpectedReturnCodes:
          PackageInUse: 1
          InstallInProgress: 2
```
* Updated manifest parsing and validation for the new fields.
* Refactored common code for checking installer exit code from MSI and ShellExecute installs into a workflow task that now also checks the expected return codes.

Not in this PR: Changes for REST sources

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1421)

yao-msft commented on schemas/JSON/manifests/v1.1.0/manifest.installer.1.1.0.json:
```diff
@@ -132,18 +132,50 @@
         }
       }
     },
+    "ReturnCode": {
```
>ReturnCode [](http://example.com/codeflow?start=5&length=10)

nit: InstallerReturnCode

JohnMcPMS commented on schemas/JSON/manifests/v1.1.0/manifest.singleton.1.1.0.json:
```diff
@@ -165,18 +165,52 @@
+    "ExpectedReturnCodes": {
+      "type": "object",
+      "properties": {
+        "PackageInUse": {
```
RebootRequired?

JohnMcPMS commented on src/AppInstallerCommonCore/Public/AppInstallerErrors.h:
```diff
@@ -80,6 +80,15 @@
 #define APPINSTALLER_CLI_ERROR_LICENSE_NOT_ACCEPTED             ((HRESULT)0x8a150041)
 #define APPINSTALLER_CLI_ERROR_PROMPT_INPUT_ERROR               ((HRESULT)0x8a150042)
 #define APPINSTALLER_CLI_ERROR_INVALID_MSIEXEC_ARGUMENT         ((HRESULT)0x8a150043)
+#define APPINSTALLER_CLI_ERROR_INSTALL_PACKAGE_IN_USE           ((HRESULT)0x8a150044)
```
I wonder if we shouldn't jump to some higher range here to intentionally keep these separate from the rest and give us the ability to add more on the future.

JohnMcPMS commented on src/AppInstallerCommonCore/Public/AppInstallerLanguageUtilities.h:
```diff
@@ -138,7 +138,7 @@ namespace AppInstaller
         {
             auto itr = m_data.find(e);
-            THROW_HR_IF_MSG(E_NOT_SET, itr == m_data.cend(), "GetVariant(%d)", e);
+            THROW_HR_IF_MSG(E_NOT_SET, itr == m_data.cend(), "GetVariant(%d)", static_cast<int>(e));
```
@sreadingMSFT Apparently you were not the only one to see it even if it didn't give me any issues.

sreadingMSFT replied:
Great. Glad the fix was easy enough, the build error was not particularly helpful.

yao-msft replied:
Ashwini and I both saw this yesterday after we merge with latest, we fixed it the same way. Looks like some new code called into this const path.

JohnMcPMS commented on src/AppInstallerCLICore/Workflows/InstallFlow.cpp:
```diff
@@ -427,15 +462,17 @@ namespace AppInstaller::CLI::Workflow
             RenameDownloadedInstaller <<
-            ShellExecuteInstallImpl;
+            ShellExecuteInstallImpl <<
+            ReportInstallerResult("ShellExecute"sv);
```
I think that MSIX should also be normalized to flow through this path.

JohnMcPMS reviewed (commented):
We need to make sure we have the full set of cases covered; "reboot required" being missing suggests that we have some to get still.

Also, the arguably more important part of this change is having a healthy set of default values per installer type and thorough thought put into how those would be overridden by the manifest.  Requiring every MSI to output the fairly standardized set of values is not helpful.

This also plays into my comment on MSIX, as we can normalize the HRESULT returns from deployment into the set of values that we have here to make all package types easier to work with in a uniform fashion.

lechacon commented:
Updated with default values per installer type.
I changed the way this is represented in the manifest because I hadn't considered multiple return codes mapping to the same response. The format is now
```
      ExpectedReturnCodes:
          - InstallerReturnCode: 1
            ReturnResponse: PackageInUse
```

yao-msft commented on schemas/JSON/manifests/v1.1.0/manifest.installer.1.1.0.json:
```diff
@@ -132,20 +132,56 @@
+        "enum": [ 0 ]
+      },
+      "minimum": -2147483648,
+      "maximum": 429496729,
```
>429496729 [](http://example.com/codeflow?start=17&length=9)

Did you accidentally change the value, the previous 4294967295 is the max for uint32

lechacon replied:
Yes, it was accidental. I'm not even sure how I managed to get a copy-paste wrong 😆

yao-msft commented on schemas/JSON/manifests/v1.1.0/manifest.installer.1.1.0.json:
```diff
@@ -132,20 +132,56 @@
+            "$ref": "#/definitions/InstallerReturnCode"
+          },
+          "ReturnResponse": {
+            "type": [ "string", "null" ],
```
>"null" [](http://example.com/codeflow?start=32&length=6)

This probably can not be null since we don't have defaults for them, let's remove the null from type

yao-msft commented on schemas/JSON/manifests/v1.1.0/manifest.installer.1.1.0.json:
```diff
@@ -132,20 +132,56 @@
+          "ReturnResponse": {
+            "type": [ "string", "null" ],
+            "enum": [
+              "PackageInUse",
```
>PackageInUse [](http://example.com/codeflow?start=15&length=12)

nit: our enum values use camel case throughout the winget manifest

yao-msft commented on src/AppInstallerCLICore/Workflows/InstallFlow.cpp:
```diff
@@ -468,11 +519,9 @@ namespace AppInstaller::CLI::Workflow
-            context.Reporter.Error() << GetUserPresentableMessage(re) << std::endl;
-            AICLI_TERMINATE_CONTEXT(re.GetErrorCode());
+            context.Add<Execution::Data::InstallerReturnCode>(re.GetErrorCode());
+            context << ReportInstallerResult("MSIX", re.GetErrorCode(), /* isHResult */ true);
```
>" [](http://example.com/codeflow?start=50&length=1)

nit: sv to match above?

yao-msft commented on src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:
```diff
@@ -1097,4 +1097,58 @@ Do you agree to the terms?</value>
+    <value>Another installation is already in progress. Try again later.</value>
+  </data>
+  <data name="InstallFlowReturnCodeFileInUse" xml:space="preserve">
+    <value>One or more file is being used. Exit the application then try again.</value>
```
>One or more file is [](http://example.com/codeflow?start=11&length=19)

nit: One or more files are ...

yao-msft commented on src/AppInstallerRepositoryCore/Rest/Schema/1_1/Json/ManifestDeserializer_1_1.cpp:
```diff
@@ -123,6 +126,27 @@ namespace AppInstaller::Repository::Rest::Schema::V1_1::Json
+                for (auto& returnCodeNode : expectedReturnCodesNode.value().get())
+                {
+                    ExpectedReturnCodeEnum returnResponse = Manifest::ConvertToExpectedReturnCodeEnum(JsonHelper::GetRawStringValueFromJsonNode(returnCodeNode, JsonHelper::GetUtilityString(ReturnResponse)).value_or(""));
+                    DWORD installerReturnCode = JsonHelper::GetRawIntValueFromJsonNode(returnCodeNode, JsonHelper::GetUtilityString(InstallerReturnCode)).value_or(0);
```
>JsonHelper::GetRawIntValueFromJsonNode [](http://example.com/codeflow?start=48&length=38)

Code analysis did not complain about converting from signed to unsigned? I think we need a static cast here

yao-msft commented on schemas/JSON/manifests/v1.1.0/manifest.singleton.1.1.0.json:
```diff
@@ -165,20 +165,57 @@
+          }
+        }
+      },
+      "maxItems": 16,
```
>16, [](http://example.com/codeflow?start=18&length=3)

Since we allow many to one mapping now, shall we increase the maxItems? I'd think 128

Also, please make the definition in this file and the other file consistent, the installer schema did not have maxItems

yao-msft commented on src/AppInstallerCommonCore/Public/AppInstallerErrors.h:
```diff
@@ -79,20 +79,38 @@
+#define APPINSTALLER_CLI_ERROR_INSTALL_CANCELLED_BY_USER        ((HRESULT)0x8a150112)
+#define APPINSTALLER_CLI_ERROR_INSTALL_ALREADY_INSTALLED        ((HRESULT)0x8a150113)
+#define APPINSTALLER_CLI_ERROR_INSTALL_DOWNGRADE                ((HRESULT)0x8a150114)
+#define APPINSTALLER_CLI_ERROR_INSTALL_BLOCKED_BY_POLICY        ((HRESULT)0x8a150115)
```
>APPINSTALLER_CLI_ERROR_INSTALL_BLOCKED_BY_POLICY [](http://example.com/codeflow?start=8&length=48)

nit: these error codes could be added to e2etests error list for completeness

yao-msft commented on src/AppInstallerCLICore/Workflows/InstallFlow.cpp:
```diff
@@ -432,15 +481,17 @@ namespace AppInstaller::CLI::Workflow
             RenameDownloadedInstaller <<
-            DirectMSIInstallImpl;
+            DirectMSIInstallImpl <<
+            ReportInstallerResult("MsiInstallProduct"sv, APPINSTALLER_CLI_ERROR_MSI_INSTALL_FAILED, /* isHResult */ true);
```
>/* isHResult */ true [](http://example.com/codeflow?start=100&length=20)

MsiInstallProduct returns win32 error codes(like 1603), not hresults. So this should be false.

yao-msft commented on src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp:
```diff
@@ -282,6 +282,74 @@ namespace AppInstaller::Manifest
+    {
+        ExpectedReturnCodeEnum result = ExpectedReturnCodeEnum::Unknown;
+
+        if (Utility::CaseInsensitiveEquals(in, "PackageInUse"))
```
>Utility::CaseInsensitiveEquals [](http://example.com/codeflow?start=12&length=30)

nit: since there're more than a couple strings, we might want to do toLower than do direct compare, like what we did for InstallerTypeEnum

yao-msft commented on src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp:
```diff
@@ -410,4 +478,48 @@ namespace AppInstaller::Manifest
+            // See https://docs.microsoft.com/en-us/windows/win32/appxpkg/troubleshooting
+            return
+            {
+                { ERROR_INSTALL_PREREQUISITE_FAILED, ExpectedReturnCodeEnum::MissingDependency },
```
>ERROR_INSTALL_PREREQUISITE_FAILED [](http://example.com/codeflow?start=18&length=33)

For MSIX apis, the returned error codes are HResults, so you'll have to convert these win 32 error codes to HResults,
otherwise, they may not get matched. Use below in allmsix return codes
HRESULT_FROM_WIN32(ERROR_INSTALL_PREREQUISITE_FAILED)

yao-msft commented on src/AppInstallerCommonCore/Manifest/ManifestValidation.cpp:
```diff
@@ -162,6 +162,20 @@ namespace AppInstaller::Manifest
+            {
+                if (!returnCodeSet.insert(code.first).second)
+                {
+                    resultErrors.emplace_back(ManifestError::DuplicateReturnCodeEntry);
```
>DuplicateReturnCodeEntry [](http://example.com/codeflow?start=61&length=24)

A negative test case for semantic checks would be ideal

yao-msft commented on src/AppInstallerCommonCore/Manifest/ManifestYamlPopulator.cpp:
```diff
@@ -625,6 +639,31 @@ namespace AppInstaller::Manifest
         return resultErrors;
     }
 
+    ValidationErrors ManifestYamlPopulator::ProcessExpectedReturnCodesNode(const YAML::Node& returnCodesNode)
```
>ProcessExpectedReturnCodesNode [](http://example.com/codeflow?start=44&length=30)

As of convention and completeness, please add the new manifest entries in MultiFileManifestV1_1 and ManifestV1_1-Singleton.yaml and update the VerifyManifestContent in unit tests.

yao-msft commented on src/AppInstallerCLIE2ETests/Constants.cs:
```diff
@@ -134,6 +134,25 @@ public class ErrorCode
             public const int ERROR_CUSTOMHEADER_EXCEEDS_MAXLENGTH = unchecked((int)0x8a150047);
+            public const int ERROR_MISSING_RESOURCE_FILE = unchecked((int)0x8a150048);
+            public const int ERROR_MSI_INSTALL_FAILED = unchecked((int)0x8a150049);
+            public const int ERROR_INVALID_MSIEXEC_ARGUMENT = unchecked((int)0x8a150050);
```
>50 [](http://example.com/codeflow?start=85&length=2)

just noticed this should be 4A

yao-msft commented on src/AppInstallerCLIE2ETests/Constants.cs:
```diff
@@ -134,6 +134,25 @@ public class ErrorCode
+            public const int ERROR_INSTALL_NO_NETWORK = unchecked((int)0x8a150107);
+            public const int ERROR_INSTALL_CONTACT_SUPPORT = unchecked((int)0x8a150108);
+            public const int ERROR_INSTALL_REBOOT_REQUIRED_TO_FINISH = unchecked((int)0x8a150109);
+            public const int ERROR_INSTALL_REBOOT_REQUIRED_TO_INSTALL = unchecked((int)0x8a150110);
```
>10 [](http://example.com/codeflow?start=95&length=2)

0A, 0B ...

yao-msft commented on src/AppInstallerCLITests/AppInstallerCLITests.vcxproj:
```diff
@@ -337,6 +340,12 @@
     <CopyFileToFolders Include="TestData\Manifest-Bad-DuplicateKey.yaml">
       <DeploymentContent>true</DeploymentContent>
     </CopyFileToFolders>
+    <CopyFileToFolders Include="TestData\Manifest-Bad-DuplicateReturnCode-ExpectedCodes.yaml">
```
>Manifest-Bad-DuplicateReturnCode-ExpectedCodes.yaml [](http://example.com/codeflow?start=41&length=51)

Looks like the new test files forgot tobe included in the commit causing test failure

yao-msft reviewed (approved):
:shipit:
